### DISCUSSION TITLE
Best Configuration to optimize battery life

### DISCUSSION THREAD

**Discussion by xselvasx** [pvvx/ATC_MiThermometer#463] [Q&A]
Hi everyone, I'm starting to use the LYWSD03MMC sensors with the custom firmware and I wanted to ask if with your experience you can tell me how to configure the various parameters of the 4.6 firmware to optimize battery life.
I am referring to the following parameters:
Sensor in "LowPower mode"
Advertising type
AdFlags
Advertising interval
Measure interval
I'm using these devices with home assistant through a bluetooh proxy

**Comment by AleXSR700** [comment]
I would also be interested and would like to add the question how much difference the RX/TX power makes. According to docs, increasing RX power needs a strobger power supply. So how much can you save by decreasing RX/TX power?

**Comment by pvvx** [comment]
Maximum battery saving:
![image](https://github.com/user-attachments/assets/d1f5bf74-ee00-46d5-af4d-8015296a6eb6)

> So how much can you save by decreasing RX/TX power?

A 50% discharged CR2032 cannot deliver more than +3 dBm of current for RF TX.
+10 dBm -> peak current 25 mA
+0 dBm -> peak current 6..7 mA

The communication distance in "LE Long Range" mode and TX 0 dBm is more than 500 meters.

**Reply by AleXSR700** [reply]
Thank you for your response. 

Would decreasing RX power below 0 dB have a similarly big impact?

PS:
Why is measurement interval always reset to 2 when I set advertising to 10000?

Advertising interval: 10000 ms, step: 62.5 ms

Measure interval:  2x(Advertising interval) = 20.0 sec

I thought I works set advertising to "x" of 10s despite the connection issue but I world guess that measurement interval has little impact on battery life so I would like to measure for every advertisement.

**Reply by pvvx** [reply]
The theoretical calculation is based on actual measurements (LYWSD03MMC B1.4), assuming that CR2032 has a nominal capacity of 200 mAh:
https://github.com/pvvx/ATC_MiThermometer?tab=readme-ov-file#average-power-consumption
Calculation in the presence of a capacitor in the power wire, which Xiaomi does not install on the printed circuit board due to greed and so that the battery gives a maximum of 40%.

**Reply by marvinlehmann** [reply]
I thought **Mijia** had a [higher consumption](https://github.com/pvvx/ATC_MiThermometer/issues/204#issuecomment-2351808789) than **BTHome v2**? And I'd have guessed that your **PVVX (Custom)** is the most efficient one or is the choice of the advertising format/type just that insignificant?

In case the choice doesn't matter, are there even any advantages for using the Passive BLE Monitor over the default Bluetooth integration (with _experimental_ passive scanning) + BTHome integration? Or does the BTHome format have any drawbacks I don't know of?

**Comment by brunoe90** [comment]
What does the "AdsFlag" checkbox do? What happend if I disable it? Would I disable Bluetooth Ads?

**Reply by pvvx** [reply]
Then in the Bluetooth integration in HomeAssistant there will be no reception during passive scanning. Also, the thermometer will not be displayed in the BT system menus in Windows, but will be available in TelinkMiFlasher and other programs.
Reception of advertising without "Flags" will work in the "Passive BLE monitor" integration and in all other normal programs. The exception is the Bluez API in Linux, since it cannot work with default values.
Bluez has all sorts of nonsense written in it and the parameter passive or active scanning affects many things. For example, the consumption of battery devices and the constant rewriting of files on the disk of the polled during active scanning devices...
That is, with default settings, Bluez tries to quickly destroy the resource of your storage device and empty the batteries of all BLE devices. Each time a beacon is received from a device during an active scan, Linux completely overwrites the device's on-disk parameters file and queries the device for additional information. During passive scanning these actions are not performed. But then the integration of Bluettoh requires the presence of transmission in the beacon of flags. This means Home Assistant thinks that you're better off replacing batteries and storage devices more often than changing program code. (The situation is similar with Zigbee in ZHA and Z2M.)

Flag transmission requires additional energy. Insignificantly, but each transmission will be 3 bytes larger and the thermometer will be available for display in all system menus of operating systems, so the neighbor will more often connect to your thermometer or require transmission of additional information during active scanning, which will also require battery power. And the zero value of the flag (disabled) indicates that this transmission from the device is not for everyone, but only for those who need it and who want to work with the thermometer.

Many original Xiaomi BLE devices transmit events without these flags, which does not prevent them from working. But for Bluetooth integration in HomeAssistant, this may cause problems.

**Comment by pyrodex** [comment]
I have a bunch of these and once they are custom flashed is there any way except the browser to adjust settings or upgrade firmware?  Would be neat if there was a CLI way to address and update/adjust the settings.

**Reply by chinswain** [reply]
Yes, you can make changes via nRF Connect (Android App) or [via an ESP32](https://github.com/pvvx/ATC_MiThermometer/discussions/678)

**Comment by chinswain** [comment]
For hardware changes, you could upgrade the battery to a CR2450:
620mAh vs 220mAh
Someone made an adaptor for it if you don't want to cut the casing down:
https://makerworld.com/en/models/959075-adapter-battery-cr2450-xiaomi-mi-2-thermometer

Remove the LCD driver chip:
![467049326-b4f1372b-5e95-481e-a05a-ec77bcec4e74](https://github.com/user-attachments/assets/23f9debe-3207-414a-90e4-f84dcf2375e7)

Add power capacitor:
<img width="151" height="163" alt="105704705-423fc600-5f20-11eb-88ca-a2db712d0222-1" src="https://github.com/user-attachments/assets/2299f203-b0af-44c0-a284-f7308e1efcae" />

I believe a [220uF 6.3V tantalum capacitor](https://github.com/pvvx/ATC_MiThermometer/issues/11#issuecomment-1804379189) is required.

I'm wondering if this would work as suggested above: 
[220uF 4V 1.1Ω@100kHz ±20% 1411 Tantalum Capacitor](https://lcsc.com/product-detail/Tantalum-Capacitors_VISHAY-TMCMB0G227MTRF_C1954120.html?s_z=n_C1954120)

Has anyone added the capacitor and tested yet? It's been discusted a few times but I don't think any long term tests.

Edit: Pvvx recommended "Installing the missing 22..47 μF capacitor (tantalum or ceramic) in the LYWSD03MMC in the power supply increases the operating time of CR2032 by 40%."
I think he means adding a low ESR 22 and 47μF capacitor to C24 and C25?

**Comment by chinswain** [comment]
I've just ordered some "47uF 10V X5R 20% SMD 1206 MLCC SAMSUNG MLCC CL31A476MPHNNNE" to test...

**Reply by chinswain** [reply]
I've applied the above capacitors:

![PXL_20250729_133328443](https://github.com/user-attachments/assets/ce061ccf-838a-4fc9-bb1f-e4fc46d0770d)

**Comment by chinswain** [comment]
47uF didn't seem enough for LE Long Range when the battery is low so I've added a TAIYO YUDEN 16V 100uF X5R 1210 20% MLCC 

![PXL_20250808_182428857](https://github.com/user-attachments/assets/14cd2304-5c78-4618-92c8-9bc8c9da210f)

**Reply by and-fed** [reply]
@chinswain There is long discussion about capacitor. Might makes sense to read, you had to add extra one: https://github.com/pvvx/ATC_MiThermometer/discussions/663.

**Reply by chinswain** [reply]
Hmm, interesting read, maybe I should add 2x 100uF in parallel (I purchased 50 from mouser to avoid fakes) ? I can't find any info on the burst transmit current draw for LE Long Range, other than it being significantly higher. My battery is at 2.3V and working fine, we'll see how low it goes.

**Reply by and-fed** [reply]
I don't pretend that my info is 100%, but I think LE Long Range should have similar max pulse current. It just lasts longer to transfer more info. Everything should be same with exception for more data should be transmitted for better error correction.

**Reply by and-fed** [reply]
I did test, default settings + LE Long Range. Pulse is about same 7mA on my device, but it lasts longer and, as the result, average current is higher too (~12uA vs ~20uA). You can compare with my other tests in mentioned topic if you would like. 
<summary><summary>
<details>
<img width="2844" height="1678" alt="image" src="https://github.com/user-attachments/assets/36adf256-c9d0-4d97-9cec-258bdc86c485" />
<details>

**Reply by chinswain** [reply]
Thanks, do you think 100uF is sufficient? My sensors are inside beehives so battery life is the main concern (I have squeezed 2450 in).

**Reply by and-fed** [reply]
I don't have enough knowledge and experience to give a good recommendations. But I would say an optimal set up (if I do just for myself). I believe you did most of them already:
- enable LE Long Range only if it's really necessary
- reduce advertisement intervals
- cr2450 + remove lcd,controller
- add low ESR and low leakage capacitor at least 100uF. It's difficult to find bigger than yours MLCC, so several in parallel might be a good option. The issue with all these capacitors - per spec there is a worst case for leakage and it's high enough. But in reality - it's much lower almost in all cases. I would say even 2-3 ( in parallel) your capacitors shouldn't have >1uA leakage. For a reliable, precise result better to measure something.

**Reply by pvvx** [reply]
> Hmm, interesting read, maybe I should add 2x 100uF in parallel (I purchased 50 from mouser to avoid fakes) ? I can't find any info on the burst transmit current draw for LE Long Range, other than it being significantly higher. My battery is at 2.3V and working fine, we'll see how low it goes.

With LE longRange we get absolutely the same transmitter current, but the beacon transmission consists of 2 parts and the bit transmission time is 8 times longer. On average, the transmitter current lasts from 15..20 times longer, taking into account the transmission of the header on the main channels, and data on secondary ones with the addition of the device name transmission.

For LE Long Range, it is advisable to use thermometers with CR2450 or 2xAAA. And increase the pause between beacons by 2 times. 
This will result in similar battery consumption as when running in Legacy mode (BT4.2).

MJWSD05MMC on the CR2450 battery from the kit worked for more than 3 years in LE Long Range (5 second interval) on the proposed firmware. No hardware modifications were made.

This one (other MJWSD05MMC) worked in LE LR in the first year and continues to work BT4.2. Likewise, no hardware changes were made.
<img width="941" height="401" alt="image" src="https://github.com/user-attachments/assets/d7cdc377-9161-4944-9b70-9df878570f43" />
All firmware versions were tested on it...

The MJWSD05MMC has a more complex program and requires more power than some LYWSD03MMC variants.

**Reply by chinswain** [reply]
So 5000 Advertising interval is best?
My listener is waking once an hour and listening to both 1M and LE Range so should be able to detect them.

**Reply by and-fed** [reply]
> So 5000 Advertising interval is best?
> My listener is waking once an hour and listening to both 1M and LE Range so should be able to detect them.

Why don't use 10sec? It helps to reduce power consumption further. 
I would even try a bigger interval, but flasher UI doesn't allow. Just need to make sure you will be able to connect to device and collect info.

**Reply by and-fed** [reply]
@chinswain Added extra info to my [old post](https://github.com/pvvx/ATC_MiThermometer/discussions/663#discussioncomment-14091003) with more capacitors and soldered x2 like yours to my device.

**Reply by chinswain** [reply]
How did you solder two capacitors? Stacked vertically? I had trouble with my custom housing so still have one currently, working ok at 2.2V.

<img width="393" height="257" alt="image" src="https://github.com/user-attachments/assets/2a4f8f70-ce69-4c14-a368-7774b62a21b3" />

**Reply by and-fed** [reply]
Yes, vertically like on the image, but without a special plates, used simple solder instead. Prepared firstly the pack, after that soldered to PCB. But my sensor is another, without special place for the pack, so I also used wires to join to GND/VCC and was able to hide inside case. 
<details><summary>Img</summary>
<img width="323" height="243" alt="image" src="https://github.com/user-attachments/assets/dea3124e-a9ba-43bf-94d1-91501fde4656" />
</details>

**Comment by GAZ082** [comment]
@pvvx  hi, why the Advertising interval and Measure Interval cannot be set to higher values than 10k/24 respectively? Is it outside BT specs?

Thanks

**Reply by AleXSR700** [reply]
Maybe pvvx will answer in more detail, but he has before (could not find the link spontaneously).

https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-guides/ble/get-started/ble-device-discovery.html#advertising-interval

I think it is the BLE-Stack limitation of 10,24s, to be precise.

**Reply by pvvx** [reply]
Because 10 x 24 = 240 seconds. This is close to the limit of the 32-bit hardware counter in the chip, which counts in units of 1/16 us.

Because in Bluetooth versions prior to 5.0, all Advertising intervals have the following values:
Interval_Max: Time Range: 20 ms to 10,485.759375 s

And with Bluetooth version **5.0**, there is a special Advertising type that is not handled by many APIs:
_Periodic_Advertising_Interval_Max_: Time Range: 7.5 ms to 81.91875 s

Check out the main Bluetooth specifications
https://www.bluetooth.com/specifications/specs/

The current version is 6.1.
